src/java/org/apache/cassandra/service/accord/AccordStateCache.java

```diff
@@ -15,547 +15,442 @@
+        maybeEvictSomeNodes();
     }
 
     public long getMaxSize()
     {
         return maxSizeInBytes;
     }
 
-    @VisibleForTesting
-    public void clear()
-    {
-        head = tail = null;
-        cache.clear();
-        saveResults.clear();
-    }
-
-    @VisibleForTesting
-    public Map<Object, AsyncResult<Void>> saveResults()
-    {
-        return saveResults;
-    }
-
-    private void unlink(Node<?, ?> node)
+    private void unlink(AccordCachingState<?, ?> node)
     {
-        Node<?, ?> prev = node.prev;
-        Node<?, ?> next = node.next;
-
-        if (prev == null)
-        {
-            Preconditions.checkState(head == node, "previous is null but the head isnt the provided node!");
-            head = next;
-        }
-        else
-        {
-            prev.next = next;
-        }
-
-        if (next == null)
-        {
-            Preconditions.checkState(tail == node, "next is null but the tail isnt the provided node!");
-            tail = prev;
-        }
-        else
-        {
-            next.prev = prev;
-        }
-
-        node.prev = null;
-        node.next = null;
+        node.unlink();
         unreferenced--;
     }
 
-    private void push(Node<?, ?> node)
+    private void link(AccordCachingState<?, ?> node)
     {
-        if (head != null)
-        {
-            node.prev = null;
-            node.next = head;
-            head.prev = node;
-            head = node;
-        }
-        else
-        {
-            head = node;
-            tail = node;
-        }
+        addLast(node);
         unreferenced++;
     }
 
-    private <K, V> void updateSize(Node<K, V> node, ToLongFunction<V> estimator)
-    {
-        bytesCached += node.estimatedSizeOnHeapDelta(estimator);
-    }
-
-    // don't evict if there's an outstanding save result. If an item is evicted then reloaded
-    // before it's mutation is applied, out of date info will be loaded
-    private boolean canEvict(Node<?, ?> node)
+    @SuppressWarnings("unchecked")
+    private <K, V> void maybeUpdateSize(AccordCachingState<?, ?> node, ToLongFunction<?> estimator)
     {
-        Invariants.checkState(node.references == 0);
-        return node.state() == FAILED || !hasActiveAsyncResult(saveResults, node.key());
+        if (node.shouldUpdateSize())
+            bytesCached += ((AccordCachingState<K, V>) node).estimatedSizeOnHeapDelta((ToLongFunction<V>) estimator);
     }
 
-    private void maybeEvict()
+    /*
+     * Roughly respects LRU semantics when evicting. Might consider prioritising keeping MODIFIED nodes around
+     * for longer to maximise the chances of hitting system tables fewer times (or not at all).
+     */
+    private void maybeEvictSomeNodes()
     {
         if (bytesCached <= maxSizeInBytes)
             return;
 
-        Node<?, ?> current = tail;
-        while (current != null && bytesCached > maxSizeInBytes)
+        Iterator<AccordCachingState<?, ?>> iter = this.iterator();
```

Comment:
Can we avoid allocating an iterator here? This is called all the time. Maybe we could add a peek method and manually iterate through the linked list?

Reply:
Introduced iterator caching to `IntrusiveLinkedList`, like Agrona does in all of its single-threaded collections, here: https://github.com/apache/cassandra-accord/pull/51

Reply:
This ended up being unnecessary (see comments on the linked cassandra-accord PR).